lorawan: add channels mask configuration

JefersonFernando opened the pull request:
This PR add channels mask configuration for LoRaWAN, creating a request for LoRaMAC. This change is needed because using  LoRaWAN with ABP activation without configuring channels mask will provide no acceptable channels on data transmit, causing data loss.
Using TheThingsNetwork I notice that the device was losing a lot of Uplinks messages. Further investigation pointed that the channels masks used by my region (AU915 FSB 2) was using only the second frequency band. The lorawan subsys did not had a channels mask configuration and was using all available channels. A mismatch of this configuration implies wrong channels being selected on Uplinks and the gateway discards it.

JefersonFernando committed:
lorawan: add channels mask configuration.

Add channels mask configuration for LoRaWAN, creating a request for
 LoRaMAC. This change is needed because using LoRaWAN with ABP
activation without configuring channels mask will provide no acceptable
 channels on data transmit, causing data loss.

Signed-off-by: Jeferson Fernando <jeferson.santos@edge.ufal.br>

JordanYates commented on subsys/lorawan/Kconfig:
```diff
@@ -44,35 +44,57 @@ config LORAWAN_PUBLIC_NETWORK
 
-config LORAMAC_REGION_AS923
-	bool "Asia 923MHz Frequency band"
+choice
```
The region configuration should be returned to how it was

JefersonFernando replied:
Done.

JordanYates commented on subsys/lorawan/Kconfig:
```diff
@@ -44,35 +44,57 @@ config LORAWAN_PUBLIC_NETWORK
+
+endchoice
+
+config LORAWAN_CHANNELS_MASK_SIZE
```
This should be a enum in the lorawan header, not a kconfig value.
Multiple regions can be enabled, which this does not express well.

JefersonFernando replied:
Fixed.

JordanYates commented on subsys/lorawan/lorawan.c:
```diff
@@ -472,6 +472,26 @@ int lorawan_set_class(enum lorawan_class dev_class)
+{
+	MibRequestConfirm_t mib_req;
+
+	if ((channels_mask == NULL) || (channels_mask_size != CONFIG_LORAWAN_CHANNELS_MASK_SIZE)) {
```
If length validation is still desired, it should be done depending on the currently enabled region, not a hardcoded value.

JefersonFernando replied:
Fixed.

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -295,6 +295,16 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
 
+/**
+ * @brief Set the channels mask.
+ *
```
It would be good to add some notes here about when you would want to use this function

JefersonFernando replied:
Fixed.

JordanYates reviewed (changes requested):
CI failures need to be fixed

rodrigopex commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
+ * @param channels_mask Buffer with channels mask to be used.
+ * @param channels_mask_size Size of channels mask buffer.
+ *
+ * @return 0 if successful, negative errno code if failure
```
Add the @ret val for each return code. 
```suggestion
 * @retval 0 successful
 * @retval -EXXXX <justify>
```

JefersonFernando replied:
Fixed.

rodrigopex commented on subsys/lorawan/lorawan.c:
```diff
@@ -472,6 +495,26 @@ int lorawan_set_class(enum lorawan_class dev_class)
 	return 0;
 }
 
+int lorawan_set_channels_mask(uint16_t *channels_mask, uint8_t channels_mask_size)
```
Please add some tests to cover this function. The current test suite seems to have that untouched.

JefersonFernando replied:
Fixed.

rodrigopex commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
+ * @brief Set the channels mask.
+ *
+ * Change the default channels mask. When mask is not changed, all the channels
+ * can be used for data transmition. Some Network Servers don't use all channels,
```
```suggestion
 * can be used for data transmission. Some Network Servers don't use all the channels,
```

JefersonFernando replied:
Fixed.

rodrigopex commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
+ *
+ * @return 0 if successful, negative errno code if failure
+ */
+int lorawan_set_channels_mask(uint16_t *channels_mask, uint8_t channels_mask_size);
```
```suggestion
int lorawan_set_channels_mask(uint16_t *channels_mask, size_t channels_mask_size);
```

JefersonFernando replied:
Fixed.

JefersonFernando committed:
tests: subsys: lorawan: add channels mask configuration tests

Checks the behavior of configuration of channels mask for LoRaWAN.
Are passed wrong masks sizes as argument, NULL pointer instead channels
mask buffer and the right case to validate the expected returns.

Signed-off-by: Jeferson Fernando <jeferson.santos@edge.ufal.br>

JefersonFernando reviewed (commented):
Fixed channels mask set function returns, comments and add tests.
